consensus/state_sync/module.go

```diff
@@ -24,41 +43,140 @@ var (
+		committedBlocksChannel: make(chan uint64, committedBlocsChannelSize),
+	}
+	m.logger = logger.Global.CreateLoggerForModule(m.GetModuleName())
 
 	for _, option := range options {
 		option(m)
 	}
-
 	bus.RegisterModule(m)
 
-	m.logger = logger.Global.CreateLoggerForModule(m.GetModuleName())
-
 	return m, nil
 }
 
-// TODO(#352): implement this function
-// Start performs state sync
 func (m *stateSync) Start() error {
-	// processes and aggregates all metadata collected in metadataReceived channel,
-	// requests missing blocks starting from its current height to the aggregated metadata's maxHeight,
-	// once the requested block is received and committed by consensus module, sends the next request for the next block,
-	// when all blocks are received and committed, stops the state sync process by calling its `Stop()` function.
+	go m.metadataSyncLoop()
 	return nil
 }
 
-// TODO(#352): check if node is a valdiator, if not send Consensus_IsSyncedNonValidator event
-// Stop stops the state sync process, and sends `Consensus_IsSyncedValidator` FSM event
-func (m *stateSync) Stop() error {
+// Start a synchronous state sync process to catch up to the network
+// 1. Processes and aggregates all metadata collected in metadataReceived channel
+// 2. Requests missing blocks until the maximum seen block is retrieved
+// 3. Perform (2) one-by-one, applying and validating each block while doing so
+// 4. Once all blocks are received and committed, stop the synchronous state sync process
+func (m *stateSync) StartSynchronousStateSync() error {
+	consensusMod := m.bus.GetConsensusModule()
+	currentHeight := consensusMod.CurrentHeight()
+	nodeAddress := consensusMod.GetNodeAddress()
+	nodeAddressBz, err := hex.DecodeString(nodeAddress)
+	if err != nil {
+		return err
+	}
+
+	readCtx, err := m.GetBus().GetPersistenceModule().NewReadContext(int64(currentHeight))
+	if err != nil {
+		return err
+	}
+	defer readCtx.Release()
+
+	// Get a view into the state of the network
+	_, maxHeight := m.getAggregatedStateSyncMetadata()
+
+	// Synchronously request block requests from the current height to the aggregated metadata height
+	// Note that we are using `<=` because:
+	// - maxHeight is the max * committed * height of the network
+	// - currentHeight is the latest * committing * height of the node
+	for currentHeight <= maxHeight {
+		m.logger.Info().Msgf("Synchronous state sync is requesting block: %d, ending height: %d", currentHeight, maxHeight)
+
+		// form the get block request message
+		stateSyncGetBlockMsg := &typesCons.StateSyncMessage{
+			Message: &typesCons.StateSyncMessage_GetBlockReq{
+				GetBlockReq: &typesCons.GetBlockRequest{
+					PeerAddress: nodeAddress,
+					Height:      currentHeight,
+				},
+			},
+		}
+		anyProtoStateSyncMsg, err := anypb.New(stateSyncGetBlockMsg)
+		if err != nil {
+			return err
+		}
+
+		// Broadcast the block request
+		if err := m.GetBus().GetP2PModule().Broadcast(anyProtoStateSyncMsg); err != nil {
```

Comment:
Isn't it safe to just get the block from any node? QCs would serve to validate the fetched block. We would need the validator set at that height though

Reply:
> Isn't it safe to just get the block from any node?

Yes. it is.

What we're doing:

1. Broadcast block request to everyone
2. Get block from some nodes

> We would need the validator set at that height though

3. Validation of QCs (and other) happens during block application of the blocks retrieved above